### PR TITLE
Benchmarks: opmphm: add time benchmarks for opmphm, binary search and hsearch

### PR DESCRIPTION
**Pull request by KurtMi**
# Purpose

Add benchmarks for the OPMPHM build/search, binary search and hsearch

# Checklist

Check relevant points but please do not remove entries.
For docu fixes, spell checking, and similar nothing
needs to be checked.

- [x] commit messages are fine ("module: short statement" syntax and refer to issues)
- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [ ] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins)
- [ ] release notes are updated (doc/news/_preparation_next_release.md)

### CODE REVIEW

**Comment by markus2330** [comment]
jenkins build all please

**Comment by markus2330** [comment]
@ingwinlu Can you take a look at the failed jobs? Seems like some jobs are misconfigured (e.g. valgrind missing?)

**Comment by markus2330** [comment]
jenkins build source-package-test please

**Comment by markus2330** [comment]
jenkins build icheck please

**Comment by ingwinlu** [comment]
I only had a look at `elektra-gcc-configure-debian-stretch-minimal` and the test is now failing because we don't have the awk script masking errors in memcheck anymore (you can see it passing in earlier build runs but with the same error message.

So either keep the system minimal and remove memchecking or install valgrind.

**Comment by markus2330** [comment]
> So either keep the system minimal and remove memchecking or install valgrind.

Could you please remove the checks on the agents where the jobs fail or no valgrind is installed? (Running valgrind on the latest versions of an distribution should be enough.)

I wonder why this did not show up directly in #1882 where the changes were done.

**Comment by sanssecours** [comment]
> I wonder why this did not show up directly in #1882 where the changes were done.

I assume that is the case, since nobody called `jenkins build all` for PR #1882.

**Comment by ingwinlu** [comment]
jenkins build gcc-configure-debian-stretch-minimal please

**Comment by ingwinlu** [comment]
jenkins build gcc-configure-debian-wheezy please

**Comment by KurtMi** [comment]
Thank you for the fixes. Yeah, I missed a spot to put a ifdef.

**Comment by KurtMi** [comment]
jenkins build all please

**Comment by KurtMi** [comment]
@markus2330 Well 3 jobs are failing but the errors are not related to my changes.
* elektra-clang: Memleaks in testmod_range
* elektra-gcc-i386: run_memcheck failed as whole
* elektra-multiconfig-gcc-stable, build_type=Release,plugins=NODEP:
```
In file included from /home/jenkins/workspace/elektra-multiconfig-gcc-stable/build_type/Release/plugins/NODEP/source/tests/gtest-1.7.0/include/gtest/internal/gtest-port.h:212:0,
                 from /home/jenkins/workspace/elektra-multiconfig-gcc-stable/build_type/Release/plugins/NODEP/source/tests/gtest-1.7.0/include/gtest/internal/gtest-internal.h:40,
                 from /home/jenkins/workspace/elektra-multiconfig-gcc-stable/build_type/Release/plugins/NODEP/source/tests/gtest-1.7.0/include/gtest/gtest.h:58,
                 from /home/jenkins/workspace/elektra-multiconfig-gcc-stable/build_type/Release/plugins/NODEP/source/src/bindings/intercept/env/tests/test_getenv.cpp:10:
/usr/include/c++/6/sstream: In destructor 'std::__cxx11::basic_stringbuf<char>::~basic_stringbuf()':
/usr/include/c++/6/sstream:65:11: internal compiler error: Illegal instruction
     class basic_stringbuf : public basic_streambuf<_CharT, _Traits>
           ^~~~~~~~~~~~~~~
```
Please review my PR, but it must not be today, it is sunday :)

**Comment by ingwinlu** [comment]
jenkins build gcc-i386 please

**Comment by ingwinlu** [comment]
jenkins build multiconfig-gcc-stable please

File: benchmarks/opmphm.c [hunk 1]
```diff
@@ -1161,6 +1974,29 @@ static size_t getPower (size_t p, size_t q)
 	return result;
 }
 
+/**
+ * @brief Compares two integer.
```
Comment (markus2330) [inline benchmarks/opmphm.c]: description misleading, better to write "comparison between integers suitable as qsort callback"

File: benchmarks/opmphm.c [hunk 2]
```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
...
+				printExit ("Sanity Check Failed: hsearch can not find element");
+			}
+		}
+
+		hdestroy_r (htab);
+		elektraFree (htab);
+	}
+	// sort repeats
+	qsort (repeats, numberOfRepeats, sizeof (size_t), cmpInteger);
+	return repeats[numberOfRepeats / 2]; // take median
```
Comment (markus2330) [inline benchmarks/opmphm.c]: you should assert (and write preconditions) that numberOfRepeats is odd.

File: benchmarks/opmphm.c [hunk 3]
```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
...
+			opmphmDel (ks->opmphm);
+			ks->opmphm = NULL;
+		}
+
+		// START MEASUREMENT
+		__asm__("");
+		gettimeofday (&start, 0);
+		__asm__("");
+
+		keyFound = ksLookup (ks, keySearchFor, KDB_O_OPMPHM | KDB_O_NOCASCADING);
```
Comment (markus2330) [inline benchmarks/opmphm.c]: Isn't a single lookup too fast to be measured precisely?

File: benchmarks/opmphm.c [hunk 4]
```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
...
+		if (!hcreate_r (nI / load, htab))
+		{
+			printExit ("hcreate_r");
+		}
+
+		while ((key = ksNext (ks)))
+		{
+			e.key = (char *) keyName (key);
+			e.data = key;
+			if (!hsearch_r (e, ENTER, &ep, htab))
```
Comment (markus2330) [inline benchmarks/opmphm.c]: Here we search for every key in a keyset? What about searching for keys not in a keyset?

File: benchmarks/opmphm.c [hunk 5]
```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
...
+		{
+			printExit ("calloc");
+		}
+
+		// START MEASUREMENT
+		__asm__("");
+		gettimeofday (&start, 0);
+		__asm__("");
+
+		if (!hcreate_r (nI / load, htab))
```
Comment (markus2330) [inline benchmarks/opmphm.c]: separate measurement for creating the hash map?

File: benchmarks/opmphm.c [hunk 6]
```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
...
+	KeySet ** keySetStorage = elektraMalloc (ksPerN * sizeof (KeySet *));
+	if (!keySetStorage)
+	{
+		printExit ("malloc");
+	}
+
+	// get KeySet shapes
+	KeySetShape * keySetShapes = getKeySetShapes ();
+
+	printf ("Run Benchmark:\n");
```
Comment (markus2330) [inline benchmarks/opmphm.c]: If you know it: which one?

File: benchmarks/opmphm.c [hunk 7]
```diff
@@ -980,6 +989,787 @@ static void benchmarkMappingAllSeeds (void)
...
+			elektraRand (&actualSearchSeed);
+		}
+
+		__asm__("");
+		gettimeofday (&end, 0);
+		__asm__("");
+		// END MEASUREMENT
+
+		// sanity checks
+		if (option & KDB_O_OPMPHM)
```
Comment (markus2330) [inline benchmarks/opmphm.c]: shouldn't that be in tests? (and not in benchmarks)

File: benchmarks/opmphm.c [hunk 8]
```diff
@@ -16,13 +19,18 @@
...
 #include "benchmarks.h"
+#ifdef HAVE_HSEARCHR
+#include <search.h>
+#endif
+
 #ifdef USE_OPENMP
 #include <omp.h>
 #endif
+
+#include "../src/libs/elektra/opmphm.c"
```
Comment (markus2330) [inline benchmarks/opmphm.c]: why is this needed?

**Comment by KurtMi** [comment]
> I cannot run the benchmarks

Could it be that you feed `0` as seed to the benckmark?

**Comment by markus2330** [comment]
> Could it be that you feed 0 as seed to the benckmark?

I do not know. It never prompted for a number, I just pressed enter. As said, a description of the method is essential (how did you execute the benchmarks, how can I reproduce it).

I think you can hard-code more information (which tests to run, which seeds). This also simplifies the necessary description for reproducibility.

File: benchmarks/opmphm.c (hunk 8 above)
Comment (KurtMi) [reply benchmarks/opmphm.c]: The mapping benchmarks uses the OPMPHM directly and the OPMPHM will not be exported.

File: benchmarks/opmphm.c (hunk 7 above)
Comment (KurtMi) [reply benchmarks/opmphm.c]: Well I have such a check in the test cases, but I added them here to be sure to measure the right thing.

File: benchmarks/opmphm.c (hunk 5 above)
Comment (KurtMi) [reply benchmarks/opmphm.c]: It is planned to measure a build i.e. from nothing to full usable hash map.

File: benchmarks/opmphm.c (hunk 4 above)
Comment (KurtMi) [reply benchmarks/opmphm.c]: This is only a sanity check, a compare between hsearch search binary search search and hsearch search OPMPHM search was not planned.

File: benchmarks/opmphm.c (hunk 3 above)
Comment (KurtMi) [reply benchmarks/opmphm.c]: This single lookup builds the OPMPHM, thus it takes some time.

**Comment by KurtMi** [comment]
> I do not know. It never prompted for a number, I just pressed enter. As said, a description of the method is essential (how did you execute the benchmarks, how can I reproduce it).

Just hitting enter is not the right usage of the benchmarks :) I will write some readme.md that explains the usage properly.

> I think you can hard-code more information (which tests to run, which seeds). This also simplifies the necessary description for reproducibility.

Hard-coding the benchmarks would lead to recompile to run different benchmarks or more files in the benchmarks folder. Both seemed not be be a good option.
Hard-coding the seeds is also tricky because for some benchmarks the number of seeds used is quiet enormous.

**Comment by KurtMi** [comment]
jenkins build all please

**Comment by markus2330** [comment]
You can also write a benchmark driver (script or similar) that calls your bechmarks and feeds them with the correct seeds.

**Comment by KurtMi** [comment]
> You can also write a benchmark driver (script or similar) that executes your benchmarks and feeds them with the correct seeds.

I will think about that, but executing all benchmarks will take quite some time, so making a script that executes all could lead to the impression that they are quick.

I pushed some update, please have a look again.
P.S. elektra-clang fails due to testmod_range.

File: benchmarks/opmphm.c (hunk 2 above)
Comment (markus2330) [reply benchmarks/opmphm.c]: If they are disabled, its the problem of the person who disabled them.

But even if they are disabled, it is a great documentation.

**Comment by markus2330** [comment]
Thank you, great job!
